### ISSUE TITLE
Invariant TypeVar not honored for class or instance variable assignments

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5703]
Consider the following test case:

```python
from typing import Generic, TypeVar

class A:
    ...

class B(A):
    ...

T = TypeVar("T")

class C(Generic[T]):
    x: T


ca = C[A]()

ca.x = A()

# This should generate an error.
ca.x = B()

cb = C[B]()

# This should generate an error.
cb.x = A()
cb.x = B()
```

Here pyright generates errors in one of the two cases but not the other.

This is related to a bug in the mypy issue tracker: https://github.com/python/mypy/issues/10427.

**Comment by erictraut** [comment]
Thinking about this more... I think the current behavior is correct. For example, if you have a `list[float]`, it's legal to add an `int` instance to this.

Looking at this another way, setting the variable `x` in the above code is similar to calling a setter `def set(self, x: T)`, and this allows subtypes of type `T`.
